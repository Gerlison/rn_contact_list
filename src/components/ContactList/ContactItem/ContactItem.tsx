import React, { memo } from 'react';
import { TouchableOpacity, View } from 'react-native';
import styled from 'styled-components/native';

import Text from '../../Text';

import { Contact } from '../../../types';

import icProfile from '../../../../assets/icons/icProfile.png';
import icEdit from '../../../../assets/icons/icEdit.png';
import icDelete from '../../../../assets/icons/icDelete.png';

interface Props {
  contact: Contact;
  onEdit: () => void;
  onDelete: () => void;
}

const ContactItem = ({ contact, onDelete, onEdit }: Props): JSX.Element => (
  <S.Container testID="contact item">
    <S.Row style={{ flex: 1 }}>
      <S.Image testID='profile icon' marginRight={16} size={26} source={icProfile} />
      <S.Text testID='contact name' numberOfLines={1} size={18}>
            {contact.name}
      </S.Text>
    </S.Row>

    <S.Row>
      <TouchableOpacity testID="edit icon" onPress={onEdit}>
        <S.Image marginRight={24} size={18} source={icEdit} />
      </TouchableOpacity>

      <TouchableOpacity testID="delete icon" onPress={onDelete}>
        <S.Image size={18} source={icDelete} />
      </TouchableOpacity>
    </S.Row>
  </S.Container>
);

const S = {
  Container: styled.View`
    flex-direction: row;
    align-items: center;
    justify-content: space-between;
    padding: 22px 0px;
  `,
  Row: styled.View`
    flex-direction: row;
    align-items: center;
  `,
  Text: styled(Text)`
    flex: 1;
    padding-right: 24px;
  `,
  Image: styled.Image.attrs({
    resizeMode: 'contain',
  })<{ size: number; marginRight?: number; marginLeft?: number }>`
    height: ${({ size }) => size}px;
    width: ${({ size }) => size}px;
  `,
};

export default memo(ContactItem);
