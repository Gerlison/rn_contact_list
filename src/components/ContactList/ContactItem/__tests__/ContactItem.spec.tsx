import React from 'react';
import { fireEvent, render } from '@testing-library/react-native';

import ContactItem from '../ContactItem';

describe('ContactItem', () => {
  it('SHOULD renders correctly', () => {
    render(<ContactItem onEdit={jest.fn()} onDelete={jest.fn()} />);
  });

  it('SHOULD match snapshot', () => {
    const sut = render(<ContactItem onEdit={jest.fn()} onDelete={jest.fn()} />);
    // expect(sut).toMatchSnapshot()
  });

  it.each([['profile icon'], ['contact name'], ['edit icon'], ['delete icon']])(
    'SHOULD have a %s',
    (testID) => {
      const { queryByTestId } = render(
        <ContactItem onEdit={jest.fn()} onDelete={jest.fn()} />,
      );
      const sut = queryByTestId(testID);
      expect(sut).toBeTruthy();
    },
  );

  it.each([['edit icon'], ['delete icon']])(
    'SHOULD call function on click on %s',
    (testID) => {
      const mockFn = jest.fn();
      const { queryByTestId } = render(
        <ContactItem onEdit={mockFn} onDelete={mockFn} />,
      );
      const sut = queryByTestId(testID);
      fireEvent.press(sut);
      expect(mockFn).toHaveBeenCalled();
    },
  );
});
