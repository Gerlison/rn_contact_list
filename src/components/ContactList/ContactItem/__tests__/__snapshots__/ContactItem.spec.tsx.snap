// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ContactItem SHOULD match snapshot 1`] = `
<View
  style={
    Array [
      Object {
        "alignItems": "center",
        "flexDirection": "row",
        "justifyContent": "space-between",
        "paddingBottom": 22,
        "paddingLeft": 0,
        "paddingRight": 0,
        "paddingTop": 22,
      },
    ]
  }
  testID="contact item"
>
  <View
    style={
      Array [
        Object {
          "alignItems": "center",
          "flexDirection": "row",
        },
        Object {
          "flex": 1,
        },
      ]
    }
  >
    <Image
      marginRight={16}
      resizeMode="contain"
      size={26}
      source={1}
      style={
        Array [
          Object {
            "height": 26,
            "width": 26,
          },
        ]
      }
      testID="profile icon"
    />
    <Text
      color="#2f2f2f"
      family="Poppins-Regular"
      numberOfLines={1}
      size={18}
      style={
        Array [
          Object {
            "color": "#2f2f2f",
            "fontFamily": "Poppins-Regular",
            "fontSize": 18,
          },
          Object {
            "flexBasis": 0,
            "flexGrow": 1,
            "flexShrink": 1,
            "paddingRight": 24,
          },
        ]
      }
      testID="contact name"
    >
      Joao Alves
    </Text>
  </View>
  <View
    style={
      Array [
        Object {
          "alignItems": "center",
          "flexDirection": "row",
        },
      ]
    }
  >
    <View
      accessible={true}
      focusable={true}
      onClick={[Function]}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      style={
        Object {
          "opacity": 1,
        }
      }
      testID="edit icon"
    >
      <Image
        marginRight={24}
        resizeMode="contain"
        size={18}
        source={1}
        style={
          Array [
            Object {
              "height": 18,
              "width": 18,
            },
          ]
        }
      />
    </View>
    <View
      accessible={true}
      focusable={true}
      onClick={[Function]}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      style={
        Object {
          "opacity": 1,
        }
      }
      testID="delete icon"
    >
      <Image
        resizeMode="contain"
        size={18}
        source={1}
        style={
          Array [
            Object {
              "height": 18,
              "width": 18,
            },
          ]
        }
      />
    </View>
  </View>
</View>
`;
